pkg/gpu/probe.go

```diff
@@ -0,0 +1,204 @@
+		EditorFlag: manager.EditType | manager.EditMaxEntries | manager.EditKeyValue,
+	}
+
+	attachCfg := uprobes.AttacherConfig{
+		Rules: []*uprobes.AttachRule{
+			{
+				LibraryNameRegex: regexp.MustCompile(`libcudart\.so`),
+				Targets:          uprobes.AttachToExecutable | uprobes.AttachToSharedLibraries,
+				ProbesSelector: []manager.ProbesSelector{
+					&manager.AllOf{
+						Selectors: []manager.ProbesSelector{
+							&manager.ProbeSelector{ProbeIdentificationPair: manager.ProbeIdentificationPair{EBPFFuncName: "uprobe__cudaLaunchKernel"}},
+							&manager.ProbeSelector{ProbeIdentificationPair: manager.ProbeIdentificationPair{EBPFFuncName: "uprobe__cudaMalloc"}},
+							&manager.ProbeSelector{ProbeIdentificationPair: manager.ProbeIdentificationPair{EBPFFuncName: "uretprobe__cudaMalloc"}},
+							&manager.ProbeSelector{ProbeIdentificationPair: manager.ProbeIdentificationPair{EBPFFuncName: "uprobe__cudaStreamSynchronize"}},
+							&manager.ProbeSelector{ProbeIdentificationPair: manager.ProbeIdentificationPair{EBPFFuncName: "uprobe__cudaFree"}},
+						},
+					},
+				},
+			},
+		},
+		EbpfConfig:            cfg.Config,
+		PerformInitialScan:    true,
+		EnableDetailedLogging: true,
+	}
+
+	attacher, err := uprobes.NewUprobeAttacher(gpuAttacherName, attachCfg, mgr, nil, &uprobes.NativeBinaryInspector{})
+	if err != nil {
+		return nil, fmt.Errorf("error creating uprobes attacher: %w", err)
+	}
+
+	if err := rlimit.RemoveMemlock(); err != nil {
+		return nil, err
+	}
+
+	p := &Probe{
+		mgr:      mgr,
+		cfg:      cfg,
+		attacher: attacher,
+	}
+
+	p.startEventConsumer()
+
+	if err := mgr.InitWithOptions(buf, &opts); err != nil {
+		return nil, fmt.Errorf("failed to init manager: %w", err)
+	}
+
+	if err := mgr.Start(); err != nil {
+		return nil, fmt.Errorf("failed to start manager: %w", err)
+	}
+
+	if err := attacher.Start(); err != nil {
+		return nil, fmt.Errorf("error starting uprobes attacher: %w", err)
+	}
+
+	return p, nil
+}
+
+// Close stops the probe
+func (p *Probe) Close() {
+	if p.attacher != nil {
+		p.attacher.Stop()
+	}
+
+	_ = p.mgr.Stop(manager.CleanAll)
+
+	if p.consumer != nil {
+		p.consumer.Stop()
+	}
+}
+
+// GetAndFlush returns the GPU stats
+func (p *Probe) GetAndFlush() (*model.GPUStats, error) {
+	now, err := ddebpf.NowNanoseconds()
+	if err != nil {
+		return nil, fmt.Errorf("getting current time: %w", err)
+	}
+
+	stats := model.GPUStats{}
+	for key, handler := range p.consumer.streamHandlers {
+		currData := handler.getCurrentData(uint64(now))
+		pastData := handler.getPastData(true)
+
+		if currData != nil {
+			currData.Key = key
+			stats.CurrentData = append(stats.CurrentData, currData)
+		}
+
+		if pastData != nil {
+			pastData.Key = key
+			stats.PastData = append(stats.PastData, pastData)
+		}
+
+		if handler.processEnded {
+			delete(p.consumer.streamHandlers, key)
+		}
+	}
+
+	return &stats, nil
+}
+
+func (p *Probe) startEventConsumer() {
+	handler := ddebpf.NewRingBufferHandler(4096)
```

Comment:
Can you use an appropriately named constant to indicate that this is the data channel size, and different from the ringbuffer size itself

Reply:
Done!